molgenis-security/src/main/java/org/molgenis/security/permission/PermissionSystemServiceImpl.java

```diff
@@ -61,15 +64,36 @@ public void giveUserWriteMetaPermissions(Collection<EntityType> entityTypes)
 			return;
 		}
 
-		SecurityContext securityContext = SecurityContextHolder.getContext();
-		runAsSystem(() -> giveUserEntityPermissionsAsSystem(securityContext, entityTypes));
+		Collection<GrantedAuthority> grantedAuthorities = getGrantedAuthorities(entityTypes);
+		giveUserEntityPermissionsAsSystem(grantedAuthorities);
 	}
 
-	private void giveUserEntityPermissionsAsSystem(SecurityContext securityContext, Collection<EntityType> entityTypes)
+	private void giveUserEntityPermissionsAsSystem(Collection<GrantedAuthority> authorities)
 	{
-		Collection<GrantedAuthority> grantedAuthorities = getGrantedAuthorities(entityTypes);
-		updateUserAuthorities(securityContext, grantedAuthorities);
-		updateSecurityContext(securityContext, grantedAuthorities);
+		String currentUsername = SecurityUtils.getCurrentUsername();
+		runAsSystem(() -> updatePersistedUserAuthorities(currentUsername, authorities));
+
+		Object principal = SecurityContextHolder.getContext().getAuthentication().getPrincipal();
+		Stream<SecurityContext> securityContexts = principalSecurityContextRegistry.getSecurityContexts(principal);
+
+		Collection<? extends GrantedAuthority> reachableAuthorities = roleHierarchy.getReachableGrantedAuthorities(
+				authorities);
+		securityContexts.forEach(
+				securityContext -> updateSecurityContextAuthorities(securityContext, reachableAuthorities));
+	}
+
+	private void updateSecurityContextAuthorities(SecurityContext securityContext,
+			Collection<? extends GrantedAuthority> newAuthorities)
+	{
+		Collection<? extends GrantedAuthority> authorities = securityContext.getAuthentication().getAuthorities();
+		List<GrantedAuthority> updatedAuthorities = new ArrayList<>(authorities);
+		updatedAuthorities.addAll(newAuthorities);
+
+		Authentication authentication = securityContext.getAuthentication();
+		UsernamePasswordAuthenticationToken newToken = new UsernamePasswordAuthenticationToken(
```

Comment:
I think that means that they'll need to re-enter the 2FA code if 2FA is enabled.

Reply:
Oops, I wasn't aware of the fact that tokens could be of a different type. This fix indeed doesn't work for 2FA.